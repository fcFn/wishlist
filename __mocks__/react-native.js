const RN = require('react-native');

RN.AppState.removeEventListener = function() {};
RN.BackAndroid.removeEventListener = jest.fn();

module.exports = RN;
