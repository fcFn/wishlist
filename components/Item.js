// This component is used for the items which constitute the list

import React, {Component} from 'react';
import {Text, PanResponder, Animated} from 'react-native';

import styles from '../configs/styles';

export default class Item extends Component {
  constructor() {
    super();
    this._animationValues = {
      _color: new Animated.Value(0),
      panX: new Animated.Value(0),
    };
    this._animationValues.color = this._animationValues._color.interpolate(
      {
        inputRange: [0, 1, 2, 3, 4],
        outputRange: [
          styles.itemColors.inactive,
          styles.itemColors.checked,
          styles.itemColors.remove,
          styles.itemColors.active,
          styles.itemColors.alreadyExists,
        ],
      }
    );
    this._PanResponder = PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onPanResponderGrant: this._onPanResponderGrant.bind(this),
      onPanResponderTerminate: this._onPanResponderTerminate.bind(this),
      onPanResponderRelease: this._onPanResponderRelease.bind(this),
      onPanResponderMove: Animated.event([null,
                          {dx: this._animationValues.panX}]),
    });
  }

  _checkForLongPress() {
    this._animationValues.panX.stopAnimation((x) => {
      if (x > -5) {
        // Prevent checking item on switching to edit item scene
        this.editing = true;
        this.props.recordItemToEdit(this.props.itemText);
      }
    });
  }

  _onPanResponderGrant() {
    this._animationValues._color.setValue(3);
    this._animationValues.panX.addListener(({value: x}) => {
      if (x > 0) {
        this._animationValues.panX.setValue(0);
      }
      if (x < -100) {
        this._animationValues._color.setValue(2);
        this._toRemove = true;
      } else {
        this._animationValues._color.setValue(3);
        this._toRemove = false;
      }
    });
    this.longPressTimeout = setTimeout(
      this._checkForLongPress.bind(this), 1000);
  }

  _clearOnRelease() {
    this._animationValues.panX.removeAllListeners();
    this._animationValues.panX.setValue(0);
    if (this.props.done) {
      this._animationValues._color.setValue(1);
    } else {
      this._animationValues._color.setValue(0);
    }
    this.editing = false;
    clearTimeout(this.longPressTimeout);
  }

  _onPanResponderRelease(_, gestureState) {
    this._animationValues.panX.stopAnimation((x) => {
      // If user changed his mind about deleting an item,
      // it shouldn't be checked or unchecked
      if (x === 0 && gestureState.dx < 5 && !this.editing) {
        this.props.checkItem(this.props.itemText);
      }
    });
    this._clearOnRelease();
    if (this._toRemove) {
      this.props.removeItem(this.props.itemText);
      this._toRemove = false;
    }
  }

  _onPanResponderTerminate() {
    this._clearOnRelease();
  }

  componentWillUpdate() {
    this._animationValues._color.setValue(0);
  }

  componentDidMount() {
    // Update colors when loading the list
    this.componentDidUpdate();
  }

  componentDidUpdate() {
    if (this.props.done) {
      this._animationValues._color.setValue(1);
    }
    if (this.props.alreadyExistsFlag) {
      this._animationValues._color.setValue(4);
    }
  }

  render() {
    let itemText = this.props.done ?
        'âœ“ ' + this.props.itemText : this.props.itemText;
    let style = [
      styles.item,
      {
        backgroundColor: this._animationValues.color,
        transform: [{translateX: this._animationValues.panX}],
      },
    ];
    return <Animated.View style={style} {...this._PanResponder.panHandlers}>
        <Text style={styles.itemText}>
          {itemText}
        </Text>
      </Animated.View>;
  }
}
