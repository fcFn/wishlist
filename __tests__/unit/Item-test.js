import React from 'react';
import 'react-native';
import Item from '../../components/Item';
import renderer from 'react-test-renderer';
import _ from 'lodash';

let item;

describe('Item component', () => {
  beforeEach(() => {
    item = new Item;
    item.props = {
      itemText: 'Oranges',
      done: true,
      recordItemToEdit: jest.fn(),
      checkItem: jest.fn(),
      removeItem: jest.fn(),
    };
    _.set(item, '_animationValues._color.setValue', jest.fn());
    _.set(item, '_animationValues.panX.removeAllListeners', jest.fn());
    _.set(item, '_animationValues.panX.setValue', jest.fn());
    _.set(item, '_animationValues._color.stopAnimation', jest.fn()
          .mockImplementation((cb) => {
            cb(7);
          }));
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Item/>);
    expect(tree).toMatchSnapshot();
  });

  describe('onPanresponderGrant', () => {
    it('correctly sets the background color on touching', () => {
      _.set(item, '_animationValues.panX.addListener', jest.fn()
          .mockImplementation((cb) => {
            cb({value: -30});
          }));
      item._onPanResponderGrant();
      expect(item._animationValues._color.setValue).toBeCalledWith(3);
      expect(item._animationValues.panX.addListener).toBeCalled();
      expect(item._toRemove).toBe(false);
    });

    it('will not allow moving the item to the right', () => {
      _.set(item, '_animationValues.panX.addListener', jest.fn()
          .mockImplementation((cb) => {
            cb({value: 10});
          }));
      item._onPanResponderGrant();
      expect(item._animationValues.panX.setValue).toBeCalledWith(0);
    });

    it('will mark item for removal if it\'s moved past the threshhold', () => {
      _.set(item, '_animationValues.panX.addListener', jest.fn()
          .mockImplementation((cb) => {
            cb({value: -1000});
          }));
      item._onPanResponderGrant();
      expect(item._animationValues._color.setValue).toBeCalledWith(2);
      expect(item._toRemove).toBe(true);
    });

    it('always sets the timer for the long press', () => {
      item._onPanResponderGrant();
      expect(item.longPressTimeout).toEqual(expect.anything());
    });
  });

  describe('_clearOnRelease', () => {
    it('clears everything if the item is no longer touched', () => {
      item._clearOnRelease();
      expect(item._animationValues.panX.removeAllListeners).toBeCalled();
      expect(item._animationValues.panX.setValue).toBeCalledWith(0);
      expect(item.editing).toBe(false);
      expect(item.longPressTimeout).toBeUndefined();
    });

    it('correctly sets the background color of the item', () => {
      item._clearOnRelease();
      expect(item._animationValues._color.setValue).toBeCalledWith(1);
      item.props.done = false;
      item._clearOnRelease();
      expect(item._animationValues._color.setValue).toBeCalledWith(0);
    });
  });

  describe('_onPanResponderRelease', () => {
    it('will not check/uncheck item if the user swiped the item around', () => {
      item._animationValues.panX.stopAnimation = jest
      .fn().mockImplementation((cb) => {
        cb(0);
      });
      item.editing = false;
      item._onPanResponderRelease(null, {dx: 4});
      expect(item.props.checkItem).toBeCalledWith(item.props.itemText);
    });

    it('will send an item to removal if it\'s swiped past threshhold', () => {
      item._toRemove = true;
      let spyClearOnRlease = jest.spyOn(item, '_clearOnRelease');
      item._onPanResponderRelease(null, {dx: 0});
      expect(spyClearOnRlease).toBeCalled();
      expect(item.props.removeItem).toBeCalledWith(item.props.itemText);
      expect(item._toRemove).toBe(false);
    });
  });

  it('sets light green background if the item is checked', () => {
    item.componentDidUpdate();
    expect(item._animationValues._color.setValue).toBeCalledWith(1);
    expect(item._animationValues._color.setValue.mock.calls[1]).toBeUndefined();
  });

  it('sets dark green background if tried to add existing item', () => {
    item.props = {
      done: true,
      alreadyExistsFlag: true,
    };
    item.componentDidUpdate();
    expect(item._animationValues._color.setValue.mock.calls[1][0]).toBe(4);
  });

  it('sends the correct item for editing', () => {
    item._checkForLongPress();
    expect(item.editing).toBe(true);
    expect(item.props.recordItemToEdit).toBeCalledWith('Oranges');
  });
});
