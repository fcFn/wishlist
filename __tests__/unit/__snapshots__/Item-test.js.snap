// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component renders correctly 1`] = `
<View
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 1)",
      "borderLeftWidth": 0,
      "borderRightWidth": 0,
      "borderTopWidth": 0,
      "borderWidth": 2,
      "height": 40,
      "paddingLeft": 10,
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 25,
      }
    }
  />
</View>
`;
